pkg/controller/source/source.go

```diff
@@ -46,18 +52,113 @@ type Source interface {
 	Start(eventhandler.EventHandler, workqueue.RateLimitingInterface, ...predicate.Predicate) error
 }
 
+var _ Source = &ChannelSource{}
+
 // ChannelSource is used to provide a source of events originating outside the cluster
 // (eh.g. GitHub Webhook callback).  ChannelSource requires the user to wire the external
 // source (eh.g. http handler) to write GenericEvents to the underlying channel.
-type ChannelSource chan event.GenericEvent
+type ChannelSource struct {
+	// once ensures the event distribution goroutine will be performed only once
+	once sync.Once
+	// source is the source channel to fetch GenericEvents
+	source chan event.GenericEvent
+	// stop is to end ongoing goroutine, and close the channels
+	stop <-chan struct{}
+	// dest is the destination channels of the added event handlers
+	dest []chan event.GenericEvent
+	// destLock is to ensure the destination channels are safely added/removed
+	destLock sync.Mutex
+}
+
+var _ inject.SourceChannel = &ChannelSource{}
+
+// InjectSourceChannel is internal should be called only by the Controller.
+// It is used to inject the source event channel initialized by the ControllerManager.
+func (cs *ChannelSource) InjectSourceChannel(c chan event.GenericEvent) error {
+	if cs.source == nil {
+		cs.source = c
+	}
+
+	return nil
+}
 
-var _ Source = ChannelSource(make(chan event.GenericEvent))
+var _ inject.Stop = &ChannelSource{}
+
+// InjectStop is internal should be called only by the Controller.
+// It is used to inject the stop channel initialized by the ControllerManager.
+func (cs *ChannelSource) InjectStop(stop <-chan struct{}) error {
+	if cs.stop == nil {
+		cs.stop = stop
+	}
+
+	return nil
+}
 
 // Start implements Source and should only be called by the Controller.
-func (ks ChannelSource) Start(
+func (cs *ChannelSource) Start(
 	handler eventhandler.EventHandler,
 	queue workqueue.RateLimitingInterface,
 	prct ...predicate.Predicate) error {
+	// source should have been injected before Start was called.
+	if cs.source == nil {
+		return fmt.Errorf("must call DoSourceChannel on ChannelSource before calling Start")
+	}
+
+	// stop should have been injected before Start was called
+	if cs.stop == nil {
+		return fmt.Errorf("must call DoStop on ChannelSource before calling Start")
+	}
+
+	cs.once.Do(func() {
+		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
+		go func() {
+			for {
+				select {
+				case <-cs.stop:
+					// Close source channel.
+					// TODO: should we close the source channel here?
+					close(cs.source)
+
+					// Close destination channels
+					cs.destLock.Lock()
```

Comment:
not sure if this is a common pattern, but if we wrap this in an in-line function we can call defer unlock.

```go
func() {
cs.destLock.Lock()
defer cs.destLock.Unlock()
for _, dst := range cs.dest {
  close(dst)
}
}()


```

Reply:
Graceful vs Easy to understand
I'm fine with both

Reply:
or 

```go
case <-cs.stop:
  cs.doStop()
case evt := <-cs.source:
  cs.DoEvent(evt)
```

Reply:
that makes sense